src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs

```diff
@@ -7469,7 +7468,153 @@ .locals init (int V_0)
 }");
         }
 
-        [Fact]
+        // Similar to above but switching on a local value rather than a parameter.
+        [ConditionalFact(typeof(CoreClrOnly))]
+        public void SwitchReadOnlySpanChar_Local()
+        {
+            var source =
+@"using System;
+class C
+{
+    static void Main()
+    {
+        ReadOnlySpan<char> chars = ""string 2"";
+        var number = chars switch
+        {
+            """" => 0,
+            ""string 1"" => 1,
+            ""string 2"" => 2,
+            _ => 3,
+        };
+        Console.WriteLine(number);
+    }
+}";
+            var compilation = CreateCompilation(source, targetFramework: TargetFramework.NetCoreApp, references: new[] { Net50.SystemMemory }, options: TestOptions.ReleaseExe, parseOptions: TestOptions.RegularPreview)
+                .VerifyEmitDiagnostics();
+            CompileAndVerify(compilation, expectedOutput: @"2")
+                .VerifyIL("C.Main",
+@"{
+  // Code size       79 (0x4f)
+  .maxstack  2
+  .locals init (System.ReadOnlySpan<char> V_0, //chars
+                int V_1)
+  IL_0000:  ldstr      ""string 2""
+  IL_0005:  call       ""System.ReadOnlySpan<char> string.op_Implicit(string)""
+  IL_000a:  stloc.0
+  IL_000b:  ldloca.s   V_0
+  IL_000d:  call       ""int System.ReadOnlySpan<char>.Length.get""
+  IL_0012:  brfalse.s  IL_003a
+  IL_0014:  ldloc.0
+  IL_0015:  ldstr      ""string 1""
+  IL_001a:  call       ""System.ReadOnlySpan<char> System.MemoryExtensions.AsSpan(string)""
+  IL_001f:  call       ""bool System.MemoryExtensions.SequenceEqual<char>(System.ReadOnlySpan<char>, System.ReadOnlySpan<char>)""
```

Comment:
nit: Might it be worthwhile to produce a private implementation help that would do `AsSpan` and `SequenceEqual`, so that we don't repeat it in every arm?

Reply:
Will leave as is since that would only shrink each call site by one method call.

Reply:
Don't we expect multiple call sites though?

Reply:
Yes, this would shrink each call site by one method call. It's not clear how that might affect perf though, particularly with respect to inlining. As we discussed offline, let's leave this calling pattern as is and address any specific perf issues as issues are identified.